File: Project_Nr_1/Part_1/Sender/SenderSocket.cs
using System.Net;
using System.Net.Sockets;

namespace Sender;
public class SenderSocket
{
    private Socket _socket;
    public bool IsConected;
    
    public SenderSocket() => 
        _socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

    public void Connect(string ipAddress, int port)
    {
        _socket.BeginConnect(new IPEndPoint(IPAddress.Parse(ipAddress), port), ConnectedCallback, null);
         
        //could be refactored
        Thread.Sleep(2000);
    }

    private void ConnectedCallback(IAsyncResult asyncResult)
    {
        if (_socket.Connected)
        {
            Console.WriteLine("Sender connected to brocker");
            _socket.EndConnect(asyncResult);
        }
        else 
        {
            Console.WriteLine("Error: Sender could not connect to brocker");
        }

        IsConected = _socket.Connected;
    }
    
    public void Send(byte[] data)
    {
        try
        {
            _socket.Send(data);
        }
        catch (Exception e)
        {
           Console.WriteLine($"Could not send data {e.Message}");
        }
       
    }
        
}

File: Project_Nr_1/Part_1/Python_Sender/Sender_socket.py
import socket
import threading


class SenderSocket:
    def __init__(self):
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.is_connected = False

    def connect(self, ip_address: str, port: int):
        thread = threading.Thread(target=self._connect_callback, args=(ip_address, port))
        thread.start()

        thread.join(timeout=2)

    def _connect_callback(self, ip_address: str, port: int):
        try:
            self._socket.connect((ip_address, port))
            self.is_connected = True
            print("Sender connected to brocker")
        except OSError as e:
            self.is_connected = False
            print(f"Error: Sender could not connect to brocker ({e})")

    def send(self, data: bytes):
        try:
            self._socket.sendall(data)
        except OSError as e:
            print(f"Could not send data {e}")